src/object/srv_obj.c

```diff
@@ -2019,17 +2019,19 @@ obj_local_rw(crt_rpc_t *rpc, struct obj_io_context *ioc, struct dtx_handle *dth)
 again:
 	rc = obj_local_rw_internal_wrap(rpc, ioc, dth);
 	if (dth != NULL && obj_dtx_need_refresh(dth, rc)) {
-		if (unlikely(++retry % 10 == 3)) {
+		if (unlikely(++retry % 10 == 9)) {
 			dsp = d_list_entry(dth->dth_share_tbd_list.next, struct dtx_share_peer,
 					   dsp_link);
 			D_WARN("DTX refresh for "DF_DTI" because of "DF_DTI" (%d) for %d times, "
-			       "maybe dead loop\n", DP_DTI(&dth->dth_xid), DP_DTI(&dsp->dsp_xid),
+			       "maybe starve\n", DP_DTI(&dth->dth_xid), DP_DTI(&dsp->dsp_xid),
 			       dth->dth_share_tbd_count, retry);
 		}
 
-		rc = dtx_refresh(dth, ioc->ioc_coc);
-		if (rc == -DER_AGAIN)
-			goto again;
+		if (!obj_rpc_is_fetch(rpc) || retry < 30) {
```

Comment:
if retry > 30 times for fetch, -DER_INPROGRESS will return to client?
then client will retry again right? do you think it is better than always call refresh here?

Reply:
Yes, if `dtx_refresh` failed so frequently, we will stop the server-side retry and return -DER_INPROGRESS to client. Not sure whether it is better, but It will at least reduce the possibility of too much ULTs caused system busy.